packages/kbn-expandable-flyout/src/hooks/useUrlState.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useCallback, useMemo } from 'react';
+import { useHistory, useLocation } from 'react-router-dom';
+import { encode, decode, RisonValue } from '@kbn/rison';
+
+const ROOT_KEY = 'experimentalUrlState' as const;
+
+const flywheel = {
+  [ROOT_KEY]: {
+    value: {} as Record<string, RisonValue | undefined>,
+    timeout: 0,
+  },
+};
+
+export const useUrlState = (key: string) => {
+  const location = useLocation();
+  const history = useHistory();
+
+  const value = useMemo(() => {
+    const searchParams = new URLSearchParams(location.search);
+    const param = searchParams.get(ROOT_KEY);
+    const decodedState = param ? decode(param) : ({} as Record<string, RisonValue>);
+    const decodedValue = (decodedState as Record<string, RisonValue> | undefined)?.[key];
+    flywheel[ROOT_KEY].value[key] = decodedValue;
+
+    return decodedValue;
+  }, [key, location.search]);
+
+  const setValue = useCallback(
+    (updatedValue: RisonValue | undefined) => {
+      flywheel[ROOT_KEY].value[key] = updatedValue;
+
+      if (flywheel[ROOT_KEY].timeout) {
+        window.clearTimeout(flywheel[ROOT_KEY].timeout);
+      }
+
+      // This batches updates to the URL state to avoid excessive history entries
+      flywheel[ROOT_KEY].timeout = window.setTimeout(() => {
```

Comment:
I worry about race conditions unless the whole application uses this implementation

Reply:
I will probably just drop this entirely